src/Hal/Component/OOP/Resolver/TypeResolver.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * (c) Jean-François Lépine <https://twitter.com/Halleck45>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Hal\Component\OOP\Resolver;
+
+
+/**
+ * Type resolver
+ *
+ * @author Jean-François Lépine <https://twitter.com/Halleck45>
+ */
+class TypeResolver
+{
+
+    const TYPE_VOID = 'void';
+    const TYPE_STRING = 'string';
+    const TYPE_INTEGER = 'integer';
+    const TYPE_FLOAT = 'float';
+    const TYPE_ARRAY = 'array';
+    const TYPE_NULL = 'null';
+    const TYPE_UNKNWON = 'unknown';
+    const TYPE_FLUENT_INTERFACE = 'fluent';
+
+    /**
+     * Resolves type of given string
+     *
+     * @param $string
+     * @return string
+     */
+    public function resolve($string)
+    {
+        $string = $cased = trim($string, ";\t\n\r\0\x0B");
+        $string = strtolower($string);
+
+        if(strlen($string) == 0) {
+            return self::TYPE_VOID;
+        }
+
+        if(preg_match('!^\d+$!', $string)) {
+            return self::TYPE_INTEGER;
+        }
+
+        if(preg_match('!^\d+\.\d+$!', $string)) {
+            return self::TYPE_FLOAT;
+        }
+
+        if('null' == $string) {
+            return self::TYPE_NULL;
+        }
+
+        if(preg_match('!(^\[|^array\()!', $string)) {
+            return self::TYPE_ARRAY;
+        }
+
+        if(preg_match('!(^\[|^array\()!', $string)) {
+            return self::TYPE_ARRAY;
+        }
+
+        if(preg_match('!^new\s*(.*)!', $string, $matches)) {
+            return trim(preg_replace('!^(new\s+)!', '', $cased));
```

Comment:
this is broken. ``return new TypeResolved($arg);`` will return the wrong stuff (probably ``TypeResolved($arg)`` in this case).

And this needs to support anonymous classes too:

```php
return new class implements Countable, Iterator {
    // ...
};
```

should be reported based on the interfaces implemented by the anonymous class.

If there are none, it should be reported as ``object`` (even though this may be less common)